File: python/proverb/proverb.py
def proverb(*args, qualifier=""):
    wordlist = [*args]
    out = []
    if not len(wordlist): return out
    first = wordlist[0]
    one = wordlist.pop(0)
    if len(wordlist):
        two = wordlist.pop(0)
        while 1:
            out.append(f"For want of a {one} the {two} was lost.")
            if not len(wordlist): break
            one = two
            two = wordlist.pop(0)
    if not qualifier: qualifier = ""
    else: qualifier += ' '
    out.append(f"And all for the want of a {qualifier}{first}.")
    return out

File: ruby/proverb/proverb.rb
class Proverb
  def initialize(*things, qualifier:nil)
    @things = things
    @qualifier = qualifier
  end

  def to_s
    @things.each_with_index.each_with_object([]) { |(thing, i), out|
      out << if i != @things.length - 1
               "For want of a #{@things[i]} the #{@things[i+1]} was lost."
             else
               "And all for the want of a #{@qualifier ? @qualifier + " " : ""}#{@things[0]}."
             end
    }.join("\n")
  end
end
